### ISSUE TITLE
Pulling Singularity image doesn’t work if SLURM node has no internet

### ISSUE THREAD

**Issue by wtraylor** [getpopper/popper#919]
On the computer cluster I am using ([Goethe HLR](https://csc.uni-frankfurt.de/)), only the login node has internet access, not the computing nodes. Therefore, building or pulling a Singularity image on the computing node does not work.

I run the workflow like this:

```sh
popper run --engine singularity --config config.yml --file wf.yml
```

The details of the workflow are not relevant, but what’s important is that Popper (version `2020.09.01`) now tries to execute `singularity pull` through SLURM, using `srun`. That happens in [src/popper/runner_slurm.py](https://github.com/getpopper/popper/blob/7fd0e0a507905b73cf5ee1b12b0e4a03671d9bc7/src/popper/runner_slurm.py#L178). Apparently, this behavior was introduced in pull request #912. 

I don’t have a good suggestion. It seems like some people want their Singularity images built on the computing node, and others (like me) on the login node.

**Comment by JayjeetAtGithub** [comment]
Thanks for opening this issue @wtraylor. This helps us to understand different Slurm environments better. We can change to building images on the login node by default and give an option to use the compute node (if singularity is not present on the login node). What do you think @wtraylor and @ivotron?

**Comment by ivotron** [comment]
we could also make use of the `--skip-pull` flag to indicate the slurm runner to not pull images when it executes.

**Comment by wtraylor** [comment]
My hunch is that it is more common to prepare everything for an experiment on the login node, while the actual execution happens on the computing nodes. For example, I would compile my application on the login node. So from my perspective it would be more intuitive to also prepare the container image on the login node.

But I am not a very seasoned HPC user.

**Comment by ivotron** [comment]
> My hunch is that it is more common to prepare everything for an experiment on the login node, while the actual execution happens on the computing nodes. For example, I would compile my application on the login node. So from my perspective it would be more intuitive to also prepare the container image on the login node.
> 
> But I am not a very seasoned HPC user.

another pattern I've heard from people working in HPC scenarios is that they don't have network connectivity at all to the outside world from the slurm cluster, not even on the login node, so they need to scp images to the login node and then run from there. In these scenarios, doing `--skip-pull` and `--skip-clone` would help. Maybe we also need a `--skip-build`?

So for this issue, do we agree that we can do this:

> We can change to building images on the login node by default and give an option to use the compute node (if singularity is not present on the login node). What do you think @wtraylor and @ivotron?

 given what @wtraylor mentioned above, this would address this issue, right?

**Comment by wtraylor** [comment]
> Maybe we also need a --skip-build?
Since building an image typically involves installing software within that image, I think skipping the build would be required, too.

**Comment by ivotron** [comment]
sorry, I didn't explain well. What I had in mind was running the entire workflow first on the frontend node once, so it builds containers, but in single-node mode (i.e. just doing `popper run` with no `-r` flag). Then subsequently running `popper run --skip-pull -r slurm`. But I agree that it would be better to do that by default, and control the behavior via config file.

**Comment by JayjeetAtGithub** [comment]
Yeah, building redundantly on multiple nodes is wrong. We can make changes to have 2 modes controlled through the config : i) build on login node (like in local without `srun`) ii) build on a single compute node (controlled through `srun`). What do you think @ivotron  ?

**Comment by ivotron** [comment]
yeah, that sounds good. I'd go further and say to not implement ii) until users request it
